src/translator/nth_element.cu

```diff
@@ -284,82 +27,38 @@ public:
                 size_t maxBatchSize,
                 DeviceId deviceId)
       : deviceId_(deviceId),
-        maxBeamSize_(maxBeamSize), maxBatchSize_(maxBatchSize),
-        NUM_BLOCKS(std::min(
-            500,
-            int(maxBeamSize* MAX_VOCAB_SIZE / (2 * BLOCK_SIZE))
-                + int(maxBeamSize* MAX_VOCAB_SIZE % (2 * BLOCK_SIZE) != 0))) {
+        maxBeamSize_(maxBeamSize), maxBatchSize_(maxBatchSize) {
     // std::cerr << "NthElement::NthElement" << std::endl;
 
     cudaSetDevice(deviceId_.no);
 
-    CUDA_CHECK(cudaMalloc((void**)&d_ind, maxBatchSize * NUM_BLOCKS * sizeof(int)));
-    CUDA_CHECK(cudaMalloc((void**)&d_out, maxBatchSize * NUM_BLOCKS * sizeof(float)));
-
-    CUDA_CHECK(cudaMalloc((void**)&d_res_idx, maxBatchSize * maxBeamSize * sizeof(int)));
-    CUDA_CHECK(cudaMalloc((void**)&d_res,     maxBatchSize * maxBeamSize * sizeof(float)));
-
-    CUDA_CHECK(cudaHostAlloc((void**)&h_res,     maxBeamSize * maxBatchSize * sizeof(float), cudaHostAllocDefault));
-    CUDA_CHECK(cudaHostAlloc((void**)&h_res_idx, maxBeamSize * maxBatchSize * sizeof(int), cudaHostAllocDefault));
-
-    CUDA_CHECK(cudaMalloc((void**)&d_breakdown, maxBeamSize * sizeof(float)));
-    CUDA_CHECK(cudaMalloc((void**)&d_batchPosition, (maxBatchSize + 1) * sizeof(int)));
-    CUDA_CHECK(cudaMalloc((void**)&d_cumBeamSizes,  (maxBatchSize + 1) * sizeof(int)));
+    const int tempElts = maxBatchSize * maxBeamSize * maxBeamSize * MAX_BLOCKS_PER_BEAM;
+    CUDA_CHECK(cudaMalloc((void**)&topk_tmp_id_buf, tempElts * sizeof(IndexType)));
```

Comment:
Can we change this to use a Marian allocator?

Reply:
I don't plan on doing this right now. The original code used cudaMallocs so I kept the same pattern. I believe the Marian team plans to remove this in the future entirely and use the topk operator so topk will use the Marian allocator when that work occurs.